runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -114,6 +162,189 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
          generateCompareBranchInstruction(cg, compareOp, node, divisorReg, snippetLabel, NULL);
          }
       }
+	
+TR::Register *
+J9::ARM64::TreeEvaluator::ArrayCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   return TR::TreeEvaluator::VMarrayCheckEvaluator(node, cg);
+   }
+
+TR::Register *
+J9::ARM64::TreeEvaluator::VMarrayCheckEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *) (cg->fe());
+   TR::Register *obj1Reg, *obj2Reg, *tmp1Reg, *tmp2Reg, *cndReg;
+   TR::LabelSymbol *doneLabel, *snippetLabel;
+   TR::Instruction *gcPoint;
+   TR::Snippet *snippet;
+   TR::RegisterDependencyConditions *conditions;
+   int32_t depIndex;
+   TR::Compilation* comp = cg->comp();
+
+   obj1Reg = cg->evaluate(node->getFirstChild());
+   obj2Reg = cg->evaluate(node->getSecondChild());
+   doneLabel = generateLabelSymbol(cg);
+   conditions = createConditionsAndPopulateVSXDeps(cg, 5);
+
+   TR::addDependency(conditions, obj1Reg, TR::RealRegister::NoReg, TR_GPR, true, cg);
+   TR::addDependency(conditions, obj2Reg, TR::RealRegister::NoReg, TR_GPR, true, cg);
+   tmp1Reg = addDependency(conditions, NULL, TR::RealRegister::NoReg, TR_GPR, true, cg);
+   tmp2Reg = addDependency(conditions, NULL, TR::RealRegister::NoReg, TR_GPR, false, cg);
+   cndReg = cg->allocateRegister(TR_CCR);
+   TR::addDependency(conditions, cndReg, TR::RealRegister::cr0, TR_CCR, cg);
+
+   // We have a unique snippet sharing arrangement in this code sequence.
+   // It is not generally applicable for other situations.
+   snippetLabel = NULL;
+
+   // Same array, we are done.
+   //
+   generateTrg1Src2Instruction(cg,TR::InstOpCode::ccmpw, node, cndReg, obj1Reg, obj2Reg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, doneLabel, TR::CC_EQ, cndReg);
+
+   // If we know nothing about either object, test object1 first. It has to be an array.
+   //
+   if (!node->isArrayChkPrimitiveArray1() && !node->isArrayChkReferenceArray1() && !node->isArrayChkPrimitiveArray2() && !node->isArrayChkReferenceArray2())
+      {
+#ifdef OMR_GC_COMPRESSED_POINTERS
```

Comment:
Do we care about compressed pointers at this point? Is this an optimization?

Reply:
I need Daryl's feedback in this. Initially I have tried to complete the implementation and still waiting for Daryl's feedback on this.